prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+			    uint64_t flags)
+{
+	return -FI_ENOSYS;
+}
+
+static ssize_t tcpx_send(struct fid_ep *ep, const void *buf, size_t len, void *desc,
+			 fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+	memset(&msg, 0, sizeof(msg));
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.desc = &desc;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+	msg.context = context;
+
+	return tcpx_sendmsg(ep, &msg, 0);
+}
+
+static ssize_t tcpx_sendv(struct fid_ep *ep, const struct iovec *iov, void **desc,
+			  size_t count, fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = iov;
+	msg.desc = desc;
+	msg.iov_count = count;
+	msg.addr = dest_addr;
+	msg.context = context;
+	return tcpx_sendmsg(ep, &msg, 0);
+
+}
+
+
+static ssize_t tcpx_inject(struct fid_ep *ep, const void *buf, size_t len,
+			   fi_addr_t dest_addr)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	memset(&msg, 0, sizeof(msg));
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+
+	return tcpx_sendmsg(ep, &msg, FI_INJECT | TCPX_NO_COMPLETION);
+
+}
+
+static ssize_t tcpx_senddata(struct fid_ep *ep, const void *buf, size_t len, void *desc,
+			     uint64_t data, fi_addr_t dest_addr, void *context)
+{
+	return -FI_ENOSYS;
+}
+
+static ssize_t tcpx_injectdata(struct fid_ep *ep, const void *buf, size_t len,
+			       uint64_t data, fi_addr_t dest_addr)
+{
+	return -FI_ENOSYS;
+}
+
+
+static struct fi_ops_msg tcpx_msg_ops = {
+	.size = sizeof(struct fi_ops_msg),
+	.recv = tcpx_recv,
+	.recvv = tcpx_recvv,
+	.recvmsg = tcpx_recvmsg,
+	.send = tcpx_send,
+	.sendv = tcpx_sendv,
+	.sendmsg = tcpx_sendmsg,
+	.inject = tcpx_inject,
+	.senddata = tcpx_senddata,
+	.injectdata = tcpx_injectdata,
+};
+
+static void tcpx_set_sockopts(int sock)
+{
+	int optval;
+	optval = 1;
+
+	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt reuseaddr failed\n");
+
+	if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt nodelay failed\n");
+
+}
+
+static void *tcpx_ep_connect_handler(void *data)
+{
+	struct tcpx_conn_handle *handle;
+	struct fi_eq_cm_entry *eq_entry;
+	struct tcpx_ep *tcpx_ep = (struct tcpx_ep *) data;
+
+	handle = tcpx_ep->handle;
+
+	eq_entry  = calloc(1, sizeof(*eq_entry));
+	if (!eq_entry) {
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "cannot allocate memory\n");
+	  goto err;
+	}
+
+	eq_entry->fid = &tcpx_ep->util_ep.ep_fid.fid;
+	eq_entry->info = &tcpx_ep->info;
+
+	tcpx_set_sockopts(handle->conn_fd);
+	if (connect(handle->conn_fd,
+		    (struct sockaddr *)handle->serv_addr,
+		    sizeof(*handle->serv_addr))) {
+
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "connect failed : %s\n", strerror(errno));
+	  goto err1;
+	}
+	fi_fd_nonblock(handle->conn_fd);
```

Comment:
May want to merge this with tcpx_set_sockopts() -> tcpx_setup_socket().

Reply:
This is using a blocking connect() call.  This will need to convert to non-blocking for better asynchronous processing.